lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachine.java

```diff
@@ -27,11 +31,27 @@ public State transition(String startState, String event, JourneyContext journeyC
                     String.format("Unknown state provided to state machine: %s", startState));
         }
 
+        if (currentPage != null && state instanceof BasicState basicState) {
+            if (isPageOrCriStateAndOutOfSync(basicState, currentPage)) {
+                return state;
+            } else if (basicState.getResponse() instanceof ProcessStepResponse) {
+                throw new UnknownStateException(
+                        String.format("Unknown state provided to state machine: %s", startState));
+            }
+        }
+
         State newState = state.transition(event, startState, journeyContext);
         if (newState instanceof NestedJourneyInvokeState) {
             return newState.transition(event, startState, journeyContext);
         }
 
         return newState;
     }
+
+    private boolean isPageOrCriStateAndOutOfSync(BasicState basicState, String currentPage) {
+        return basicState.getResponse() instanceof PageStepResponse pageStepResponse
+                        && !pageStepResponse.getPageId().equals(currentPage)
+                || basicState.getResponse() instanceof CriStepResponse criStepResponse
+                        && !criStepResponse.getCriId().equals(currentPage);
```

Comment:
Can `criStepResponse.getCriId().equals(currentPage)` ever be true? (i.e. can we just remove that condition)

Reply:
Ah yes I forgot we refactored the frontend to remove this.

I guess there's an argument to still validate against this in case we change the front-end logic in the future and it still would be an invalid path, but happy to remove if you think it's not necessary?

Reply:
Actually on second thoughts we probably should remove it to avoid confusion